src/memcpy_vfp.c

```diff
@@ -0,0 +1,175 @@
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "driver.h"
+
+#define BLOCK_SIZE  4096
+
+static char tmp_buf0[BLOCK_SIZE] __attribute__((aligned (128)));
+
+static inline void tegra_aligned_block_copy_vfp(void *dst, void *src, int size)
+{
+    asm volatile(
+        "   .fpu vfpv3-d16          \n\t"
+        "   .arch armv7a            \n\t"
+        "0:                         \n\t"
+        "   pld   [%1, #128]        \n\t"
+        "   pld   [%1, #160]        \n\t"
+        "   pld   [%1, #192]        \n\t"
+        "   pld   [%1, #224]        \n\t"
+        "   pld   [%1, #256]        \n\t"
+        "   subs  %2, %2, #128      \n\t"
+        "   vldm  %1!, {d0-d15}     \n\t"
+        "   vstm  %0!, {d0-d15}     \n\t"
+        "   bgt   0b                \n\t"
+        : "+r" (dst), "+r" (src), "+r" (size)
+        :
+        : "cc", "d8", "d9", "d10", "d11", "d12", "d13", "d14", "d15");
+}
+
+static void tegra_do_memcpy_vfp(char *__dst, char *__src, int size,
+                                bool buffering, bool move)
+{
+    int i, dir, block_size = BLOCK_SIZE;
+    char *tmpbuf = tmp_buf0;
+    char *src = (char *)__src;
+    char *dst = __dst;
+
+    if (0) {
+        assert(((unsigned long)src & 127) == 0);
+        assert(((unsigned long)dst & 127) == 0);
+        assert((size & 127) == 0);
+    }
+
+copy:
+    if (size < block_size)
+        block_size = size;
+
+    dir = (dst > src && move) ? -1 : 1;
+
+    if (dir < 0) {
+        src += size - block_size;
+        dst += size - block_size;
+    }
+
+    for (i = 0; i < size / block_size; i++) {
+        if (!buffering)
+            tmpbuf = dst;
+
+        tegra_aligned_block_copy_vfp(tmpbuf, src, block_size);
+
+        if (buffering)
+            tegra_aligned_block_copy_vfp(dst, tmpbuf, block_size);
+
+        src += block_size * dir;
+        dst += block_size * dir;
+    }
+
+    size -= block_size * i;
+    dst = max(dst, __dst);
+    src = max(src, __src);
+
+    if (size)
+        goto copy;
```

Comment:
Do you really need to use goto here? Seems to me like 
```c
do {
   ...
} while (size)
```
...would do the trick and be more readable...

Reply:
Okay!